# HopDropper
Remove PCR artefacts from paired-end sequencing data
## Background
PCR is an integral technique is many genomic sequencing methods. Applications include: enriching target material in nucleic acid specimens, attaching barcode indices during library preparation, amplifying material captured by hybridization probes, and generating clusters on sequencer flow cells. These PCRs introduce artefacts into sequencing data that complicate downstream analysis. HopDropper concerns itself with the type of artefact called index hops.

Index hops occur when chimeras form between molecules from different libraries, thus attaching barcode indices to material not derived from the specimen they represent. Unique dual indexing enables the identification of most index hops but cannot detect so-called “double hops”. These occur when chimeras form from templates that are themselves index hops, thereby reconstituting a valid pair of indices with exogenous library material inserted between them.

Double hops are rare, but regularly cause false positives in applications attempting detection and characterization of rare, low-abundance target material. To make matters worse, the formation of index hops chimeras is often exacerbated by target enrichment methods commonly employed in these applications, e.g. hybridization probe capture where PCR amplification is performed on low-complexity material from pooled libraries.

HopDropper removes index hops by identifying all reads that could have been derived from the same template molecule. It ensures these reads are present in only one library, discarding reads from other libraries where they are less abundant. In cases where it is not clear which library is the true source of those reads, they are removed from all libraries. HopDropper is purposefully aggressive, subscribing to the principle that confidence in results is more important than retention of suspect data. In applications concerned with the detection and characterization of rare target material, HopDropper prefers false negatives to false positives.

HopDropper identifies reads that could have been derived by the same template molecule using "intrinsic" UMIs. Intrinsic UMIs are generated by taking the first k nucleotides from a read sequence. All reads sharing the same intrinsic UMI could be copies of the same template molecule. This is premised on genomic sequences being inherently complex and input nucleic acids being randomly fragmented during library. Thus, HopDropper should not be used on amplicons!

Another limitation of HopDropper is the possibility of “UMI collisions”, i.e. multiple template molecules sharing the same UMI through coincidence, e.g. a k-mer that repeats in the genome or two template molecules that fragment at the same position. k-mers repetition can be mitigated by choosing longer k-mers (the HopDropper default is 14-mers), but the frequency of collisions from repeated k-mers will ultimately depend on the organism being sequenced, the length of its genome, and the presence of repetitive elements.

Collisions from coincidental fragmentation at the same position will depend on the length of genome as well, with shorter genomes more likely to produce collisions. Coincidental fragmentation will also become more likely as the number of genome copies in the specimens increases. On the one hand, a silver lining of rare target detection is that genome copies are necessarily limited. On the other hand, throughput and cost considerations may encourage extensive library multiplexing, thereby increasing genome copies.

Due to the aggressive nature of the HopDropper algorithm, collisions will typically result in some amount of lost data. If one of the colliding UMIs is significantly less abundant, its reads will be removed from its library while reads from the dominant colliding UMI will be retained in the other library. Alternatively, if no colliding UMI is significantly more abundant than the others, all reads will be removed from all libraries. Once again, HopDropper prefers false negatives to false positives.

Finally, intrinsic UMI collisions can be substantially mitigated using library adapters that introduce exogenous UMIs. These introduce an additional coincidence that must occur alongside k-mer repeats or fragmentation of multiple templates at the same position. If inline UMI adapters are used, the exogenous UMI sequence is already prepended to the read sequence and its intrinsic UMI. Simply run HopDropper with a longer UMI length parameter (length of exogenous UMI + length of intrinsic UMI). If adapter UMIs are instead located in barcode index regions, they must be prepended to read sequences before running HopDropper.

## The HopDropper Algorithm
1.	Enumerate UMIs from each read. This is done by taking the first k letters of the read sequence. It requires that all base calls inside the UMI have a minimum PHRED quality score, otherwise the read and its mate are discarded.
2.	For each UMI, count the number of times it occurs across all libraries. If it does not exceed a certain threshold, discard all reads containing that UMI and their mates. By default, HopDropper removes UMI singletons.
3.	For each UMI, count the number of times it appears in every library separately. Identify the library where it occurs the most times (the top library) and the library where it occurs the next most times (the second library). Determine if the number of occurrences in the top library is significantly greater than the number of occurrences in the second library. If it is significantly greater, assign the top library as that UMI’s best library. If it is not significantly greater, assign no library as that UMI’s best library.
HopDropper determines significance by whether the number of occurrences in the top library is at least twice the number of occurrences in the second library. This is premised on the doubling nature of PCR amplification and the fact that amplification of PCR artefacts will necessarily trail the templates on which they are based by at least one amplification cycle. This principle has been previously used in other chimera detection tools, including VSEARCH.
4.	Iterate through each read pair in every library. Extract UMIs from both reads. If either UMI has no best library assigned to it, discard the read pair. If the library in which the read pair is located is not the best library for both UMIs in the pair, discard the read pair.

## HopDropper usage
```
$ python hop_dropper.py -i <input file> -o <output name> -l <UMI length> -q <UMI quality> -m <minimum UMI count>
```
<b>Arguments:</b>
    -i : path to input file (described below)
    -o : name to append to output files
    -l : the number of bases to use as UMI from each read
    -q : the minimum PHRED score of each base in the UMI required for the read to be considred
    -m : the minimum number of times a UMI must appear to be considered
The HopDropper input file is a TXT file listing all FASTQ files to be analyzed. Each library must be provided as a single interleaved FASTQ file.
